File: cracking_interview/recursion/subsets.cpp
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;


vector<vector<int>> subsetsHelper(const vector<int>& nums, size_t idx) {
    if (idx == nums.size()) {
        return {{}};
    }

    auto currentSubsets = subsetsHelper(nums, idx + 1);
    auto currentSubsetsSize = currentSubsets.size();

    for (size_t i = 0; i < currentSubsetsSize; ++i) {
        currentSubsets.push_back(currentSubsets[i]);
        currentSubsets.back().push_back(nums[idx]);
    }

    return currentSubsets;
}

vector<vector<int>> subsets(const vector<int>& nums) {
    return subsetsHelper(nums, 0);
}

vector<vector<int>> subsetsIterative(const vector<int>& nums) {
    vector<vector<int>> result;

    for (size_t i = 0; i < 1 << nums.size(); ++i) {
        size_t idx = 0, j = i;
        result.emplace_back();

        while (j) {
            if (j & 1) {
                result.back().push_back(nums[idx]);
            }

            j >>= 1;
            ++idx;
        }
    }

    return result;
}

int main() {
    vector<int> nums = {1, 2, 3};

    auto subs = subsetsIterative(nums);

    for (const auto& subset: subs) {
        for (auto el: subset) {
            cout << el << ' ';
        }
        cout << endl;
    }

    return 0;
}

File: leetcode/Subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList();

        for (int i = 0; i < 1 << nums.length; ++i) {
            List<Integer> subset = new ArrayList();
            int mask = 1;
            for (int j = 0; j < nums.length; ++j) {
                if ((i & mask) != 0) {
                    subset.add(nums[j]);
                }

                mask <<= 1;
            }

            result.add(subset);
        }

        return result;
    }
}
